lib/output/switch.go

```diff
@@ -0,0 +1,341 @@
+
+// Switch is a broker that implements types.Consumer and broadcasts each message
+// out to an array of outputs.
+type Switch struct {
+	running int32
+
+	logger log.Modular
+	stats  metrics.Type
+
+	throt *throttle.Type
+
+	transactions <-chan types.Transaction
+
+	outputTsChans  []chan types.Transaction
+	outputResChans []chan types.Response
+	outputs        []types.Output
+	outputNs       []int
+	confs          []switchConfigOutputs
+	firstMatch     bool
+
+	closedChan chan struct{}
+	closeChan  chan struct{}
+}
+
+// NewSwitch creates a new FanOut type by providing outputs. Messages will be sent to
+// a subset of outputs according to condition and fallthrough settings
+func NewSwitch(
+	conf Config,
+	mgr types.Manager,
+	logger log.Modular,
+	stats metrics.Type,
+) (Type, error) {
+	lOutputs := len(conf.Switch.Outputs)
+	if lOutputs < 2 {
+		return nil, ErrSwitchNoOutputs
+	}
+
+	outputs := make([]types.Output, lOutputs)
+	confs := make([]switchConfigOutputs, lOutputs)
+	var err error
+	for i, oConf := range conf.Switch.Outputs {
+		outputs[i], err = New(oConf.Output, mgr, logger, stats)
+		if err != nil {
+			return nil, err
+		}
+		if oConf.Condition != nil {
+			oConf.cond, err = condition.New(*oConf.Condition, mgr, logger, stats)
+			if err != nil {
+				return nil, err
+			}
+			confs[i] = oConf
+		}
+	}
+
+	return newSwitch(outputs, confs, mgr, logger, stats)
+}
+
+func newSwitch(
+	outputs []types.Output,
+	confs []switchConfigOutputs,
+	mgr types.Manager,
+	logger log.Modular,
+	stats metrics.Type,
+) (*Switch, error) {
+	o := &Switch{
+		running:      1,
+		stats:        stats,
+		logger:       logger.NewModule(".broker.switch"),
+		transactions: nil,
+		outputs:      outputs,
+		outputNs:     []int{},
+		confs:        confs,
+		closedChan:   make(chan struct{}),
+		closeChan:    make(chan struct{}),
+	}
+
+	o.throt = throttle.New(throttle.OptCloseChan(o.closeChan))
+
+	o.outputTsChans = make([]chan types.Transaction, len(o.outputs))
+	o.outputResChans = make([]chan types.Response, len(o.outputs))
+	for i := range o.outputTsChans {
+		o.outputNs = append(o.outputNs, i)
+		o.outputTsChans[i] = make(chan types.Transaction)
+		o.outputResChans[i] = make(chan types.Response)
+		if err := o.outputs[i].Consume(o.outputTsChans[i]); err != nil {
+			return nil, err
+		}
+	}
+	return o, nil
+}
+
+//------------------------------------------------------------------------------
+
+// Consume assigns a new transactions channel for the broker to read.
+func (o *Switch) Consume(transactions <-chan types.Transaction) error {
+	if o.transactions != nil {
+		return types.ErrAlreadyStarted
+	}
+	o.transactions = transactions
+
+	go o.loop()
+	return nil
+}
+
+//------------------------------------------------------------------------------
+
+// loop is an internal loop that brokers incoming messages to many outputs.
+func (o *Switch) loop() {
+	defer func() {
+		for _, c := range o.outputTsChans {
+			close(c)
+		}
```

Comment:
Would be good to block here until all child outputs are closed.

Reply:
I agree. It's not clear to me how to achieve this though, without changing the behavior to call each child outputs `CloseAsync` and `WaitForClose` methods, but even with that approach, it's not clear how to pass the wait timeout duration to child outputs.. how is the `fan_out` broker accomplishing this? This output borrowed heavily from that implementation.

Reply:
Looks like the output brokers are missing this behaviour which I should look at, the dynamic fan out does it this way:

``` go
for _, ow := range d.outputs {
	ow.output.CloseAsync()
}
for _, ow := range d.outputs {
	if err := ow.output.WaitForClose(time.Second); err != nil {
		for err != nil {
			err = ow.output.WaitForClose(time.Second)
		}
	}
}
```

It's fine to use a hardcoded `time.Second` here, as the intention is to block forever regardless until it exits.